KEP-4671: Add Workload Aware Scheduling blog post for v1.36 (#54667)
kubernetes/website: content/en/blog/_posts/2026/workload-aware-scheduling-1-36.md

---
layout: blog
title: "Kubernetes v1.36: Advancing Workload-Aware Scheduling"
draft: true
date: 2026-XX-XX
slug: kubernetes-v1-36-advancing-workload-aware-scheduling
author: >
  Maciej Skoczeń (Google),
  Antoni Zawodny (Google),
  Matt Matejczyk (Google),
  Bartosz Rejman (Google),
  Jon Huhn (Microsoft),
  Maciej Wyrzuc (Google),
  TBD
---

> Scheduling large AI/ML and batch workloads is a much more complex operation than scheduling a single Pod.

Author: Good idea, done

In Kubernetes v1.35, we introduced the first tranche of *workload-aware scheduling* improvements,
featuring the foundational Workload API alongside basic *gang scheduling* support built on a Pod-based framework,
and an *opportunistic batching* feature to efficiently process identical Pods.

> Now, Kubernetes v1.36 brings a massive architectural upgrade. We have redesigned the API to cleanly separate concerns:

Author: Applied

the Workload API now serves as a static template, while the new PodGroup API describes the runtime object.
To power this, the `kube-scheduler` features a new *PodGroup scheduling cycle* that enables atomic workload processing
and paves the way for future enhancements. We are also rolling out the first iterations of *topology-aware scheduling*
and *workload-aware preemption* to advance the scheduling capabilities for these workloads. Additionally,

> we have added *ResourceClaim support for workloads* to unlock *Dynamic Resource Allocation

Reviewer: Does "we" mean:
  Maciej Skoczeń (Google),
  Antoni Zawodny (Google),
  Matt Matejczyk (Google),
  Bartosz Rejman (Google),
  Jon Huhn (Microsoft),
  Maciej Wyrzuc (Google),
  TBD
  
?

If not: reword for cvlaridy.

Author: Done

([DRA](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/))* for PodGroups. Finally,
to demonstrate real-world readiness, v1.36 delivers the first phase of integration between the Job controller and the new API.

## Workload and PodGroup API updates

The Workload API now serves as a static template, while the new PodGroup API describes the runtime object.
Kubernetes v1.36 introduces the Workload and PodGroup APIs as part of the
`scheduling.k8s.io/v1alpha2` {{< glossary_tooltip text="API group" term_id="api-group" >}},
completely replacing the previous `v1alpha1` API version.

In v1.35, Pod groups and their runtime states were embedded within the Workload resource.

> In the new model, the Workload object is rarely updated (as a template object),

Author: The spec is immutable, but in the context of this sentence, the labels and other metadata are irrelevant. Reworded this part so as not to mention updates

while the PodGroup handles the runtime state. This decoupling also improves performance and scalability,
as the PodGroup API allows per-replica sharding of status updates.

Because the Workload API acts merely as a template, the `kube-scheduler`'s logic is streamlined.
The scheduler can directly read the PodGroup, which contains all the information required by the scheduler,
without needing to watch or parse the Workload object itself.

Here is what the updated configuration looks like. First, you define the Workload object,
which now acts as a static template for your Pod groups:

```yaml
apiVersion: scheduling.k8s.io/v1alpha2
kind: Workload
metadata:
  name: training-job-workload
  namespace: some-ns
spec:
  # Pod groups are now defined as templates

>   podGroupTemplates:

Author: Just a spec, added this info to the comment

  - name: workers
    schedulingPolicy:
      gang:
        # The gang is schedulable only if 4 pods can run at once
        minCount: 4
```

> Next, workload controllers (such as the Job controller) stamp out runtime PodGroup instances based on those templates.

Reviewer 2: Actually, IIRC, we ended up not supporting this (BYOW - bring your own workload). The way it works is that Job controller automatically creates Workoad+Podgroup for eligible jobs. @helayoty to confirm

Author: Reworded this section to say the workload controllers create the Workload object

> Does this mean that if I have a CronJob I could end up having a new Workload created every _n_ minutes? (If so, might be worth calling out the limitation).

Yes, for now, a Workload+PodGroup would be created for each new Job created by CronJob (when everything is enabled in the Job controller). Maybe we could mention that in the "Integration with the Job controller" section (I'm still waiting for an author to send me that content).

The PodGroup runtime object holds the actual scheduling policy and references the template from which it was created.
It also has a status containing conditions that mirror the states of individual Pods,
reflecting the overall scheduling state of the group:

```yaml
apiVersion: scheduling.k8s.io/v1alpha2
kind: PodGroup
metadata:
  name: training-job-workers-pg
  namespace: some-ns
spec:
  # The PodGroup references the Workload template it originated from
  podGroupTemplateRef:
    workload:
      workloadName: training-job-workload

>       podGroupTemplateName: workers

Author: Added a comment. We want to bind a lifecycle of this pod group to the "true" workload, e.g., Job, not the static Workload object. Also, having a separate field, like `podGroupTemplateRef` makes the reference more verbose and easy to access.

## PodGroup scheduling cycle and gang scheduling

To efficiently manage these workloads, the kube-scheduler now features a dedicated *PodGroup scheduling cycle*.
Instead of evaluating and reserving resources sequentially Pod-by-Pod, which risks scheduling deadlocks,
the scheduler evaluates the group as a unified operation.

When the scheduler pops a PodGroup member from the scheduling queue, regardless of the group's specific policy,
it fetches the rest of the queued Pods for that group, sorts them deterministically,
and executes an atomic scheduling cycle as follows:

1. The scheduler takes a single snapshot of the cluster state to prevent race conditions and ensure consistency
   while evaluating the entire group.

2. It then attempts to find valid Node placements for all Pods in the group using a PodGroup scheduling algorithm,
   which leverages the standard Pod-based filtering and scoring phases.

3. Based on the algorithm's outcome, the scheduling decision is applied atomically for the entire PodGroup.

   * Success: If the placement is found and group constraints are met, the schedulable member Pods
     are moved directly to the binding phase together.

>    * Failure: If the group fails to meet its requirements, the entire group is considered unschedulable.

Author: You're right, added a few sentences here

This cycle acts as the foundation for *gang scheduling*. When your workload requires strict *all-or-nothing* placement,
the `gang` policy leverages this cycle to prevent partial deployments that lead to resource wastage and potential deadlocks.

While the scheduler still holds the Pods in the `PreEnqueue` until the `minCount` requirement is met, the actual scheduling phase now relies entirely
on the new PodGroup cycle. Specifically, during the algorithm's execution, the scheduler verifies
that the number of schedulable Pods satisfies the `minCount`. If the cluster cannot accommodate the required minimum,
none of the pods are bound. The group fails and waits for sufficient resources to free up.

### Limitations

The first version of the PodGroup scheduling cycle comes with certain limitations:

* For basic *homogeneous* Pod groups (i.e., those where all Pods have identical scheduling requirements
  and lack inter-Pod dependencies like affinity, anti-affinity, or topology spread constraints),
  the algorithm is expected to find a placement if one exists.

* For *heterogeneous* Pod groups, finding a valid placement is not guaranteed.

> * For Pod groups with *inter-Pod dependencies*, finding a valid placement is not guaranteed.

Reviewer: We don't guarantee finding a valid placement without PodGroups. Maybe we can reword to say that there may be valid placements but we don't yet expect the scheduler to find them, even when the solution might seem trivial or otherwise feasible to resolve.

Author: Extended these sentences

called *workload-aware preemption*. When a PodGroup cannot be scheduled, the scheduler utilizes this mechanism
to try making a scheduling of this PodGroup possible.

Compared to the default preemption used in the standard Pod-by-Pod scheduling cycle, this new mechanism
treats the entire PodGroup as a single preemptor unit. Instead of evaluating preemption victims on each Node separately,
it searches across the entire cluster. This allows the scheduler to preempt Pods from multiple Nodes simultaneously,
making enough space to schedule the whole PodGroup afterwards.

Workload-aware preemption also introduces two additional concepts directly to the PodGroup API:

* PodGroup `priority` that overrides the priority of the individual Pods forming the PodGroup.

* PodGroup `disruptionMode` that dictates whether the Pods within a PodGroup can be preempted independently,
  or if they have to be preempted together in an *all-or-nothing* fashion.

In Kubernetes v1.36, these fields are only respected by the workload-aware preemption mechanism.
We plan to extend support for these fields to other disruption sources, including default preemption

> used in the Pod-by-Pod scheduling cycle, in future releases.

Reviewer: ```suggestion
The people working on this set of features are hoping to extend support for these fields to other disruption sources, including default preemption used in the Pod-by-Pod scheduling cycle, in future releases.
```
(avoid "we" when it doesn't exactly mean "we, the people who wrote this blog post").

    - name: pg-claim
      resourceClaimTemplateName: my-claim-template
---
apiVersion: v1
kind: Pod
metadata:
  name: topology-aware-workers-pg-pod-1
spec:
  ...
  schedulingGroup:
    podGroupName: training-job-workers-pg
  resourceClaims:
    - name: pg-claim
      resourceClaimTemplateName: my-claim-template
---
apiVersion: v1
kind: Pod
metadata:
  name: topology-aware-workers-pg-pod-2
spec:
  ...
  schedulingGroup:
    podGroupName: training-job-workers-pg
  resourceClaims:
    - name: pg-claim
      resourceClaimTemplateName: my-claim-template
```

In addition, ResourceClaims referenced by PodGroups, either through
`resourceClaimName` or the claim generated from `resourceClaimTemplateName`,
become reserved for the entire PodGroup. Previously, kube-scheduler could only
list individual Pods in a ResourceClaim's `status.reservedFor` field which is
limited to 256 items. Now, a single PodGroup reference in `status.reservedFor`
can represent many more than 256 Pods, allowing high-cardinality sharing of
devices.

Together, these changes enable massive workloads with complex topologies to
utilize DRA for scalable device management.

## Integration with the Job controller

> TBD

Reviewer 3: IMO we should just remove this section.

Job controller will stay alpha in 1.37 anyway for future API changes so I think its okay to skip this update for 1.36.

WDYT @helayoty?

> TBD

Reviewer 2: I think adding a few sentences here would be beneficial, unless the job integration is already documented elsewhere. Since this is our only real controller integration so far, it's likely the easiest way for users to try out WAS.

> TBD

Reviewer 3: There is https://github.com/kubernetes/website/pull/54489.

Okay I think its fine to add then.

> TBD

Author: Added this section, PTAL

> TBD

Reviewer 4: Section has been updated and I believe it's ready now. PTAL.

---
layout: blog
title: "Kubernetes v1.36: Advancing Workload-Aware Scheduling"
draft: true
date: 2026-XX-XX
slug: kubernetes-v1-36-advancing-workload-aware-scheduling
author: >
  Maciej Skoczeń (Google),
  Antoni Zawodny (Google),
  Matt Matejczyk (Google),
  Bartosz Rejman (Google),
  Jon Huhn (Microsoft),
  Maciej Wyrzuc (Google)
---

AI/ML and batch workloads introduce unique scheduling challenges that go beyond simple Pod-by-Pod scheduling.
In Kubernetes v1.35, we introduced the first tranche of *workload-aware scheduling* improvements,
featuring the foundational Workload API alongside basic *gang scheduling* support built on a Pod-based framework,

> and an *opportunistic batching* feature to efficiently process identical Pods.

Reviewer 4: Did we manage to integrate the `opportunistic batching` for gang-scheduling plugin?

Author: Opportunistic batching is enabled by default also during gang scheduling as long as the member pods are identical

Kubernetes v1.36 introduces a significant architectural evolution by cleanly separating API concerns:
the Workload API acts as a static template, while the new PodGroup API handles the runtime state.
To support this, the `kube-scheduler` features a new *PodGroup scheduling cycle* that enables atomic workload processing
and paves the way for future enhancements. This release also debuts the first iterations of *topology-aware scheduling*
and *workload-aware preemption* to advance scheduling capabilities. Additionally,
*ResourceClaim support for workloads* unlocks *Dynamic Resource Allocation
([DRA](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/))* for PodGroups. Finally,
to demonstrate real-world readiness, v1.36 delivers the first phase of integration between the Job controller and the new API.

## Workload and PodGroup API updates

The Workload API now serves as a static template, while the new PodGroup API describes the runtime object.
Kubernetes v1.36 introduces the Workload and PodGroup APIs as part of the
`scheduling.k8s.io/v1alpha2` {{< glossary_tooltip text="API group" term_id="api-group" >}},
completely replacing the previous `v1alpha1` API version.

> In v1.35, Pod groups and their runtime states were embedded within the Workload resource.

Reviewer 4: I'd state in breif why this was not the good way to follow and why we decoupled PodGroup.

Author: This paragraph describes that:
> This separation also improves performance and scalability as the PodGroup API allows per-replica sharding of status updates.

spec:
  ...
  resourceClaims:
    - name: pg-claim
      resourceClaimTemplateName: my-claim-template
---
apiVersion: v1
kind: Pod
metadata:
  name: topology-aware-workers-pg-pod-1
spec:
  ...
  schedulingGroup:
    podGroupName: training-job-workers-pg
  resourceClaims:
    - name: pg-claim
      resourceClaimTemplateName: my-claim-template
---
apiVersion: v1
kind: Pod
metadata:
  name: topology-aware-workers-pg-pod-2
spec:
  ...
  schedulingGroup:
    podGroupName: training-job-workers-pg
  resourceClaims:
    - name: pg-claim
      resourceClaimTemplateName: my-claim-template
```

In addition, ResourceClaims referenced by PodGroups, either through
`resourceClaimName` or the claim generated from `resourceClaimTemplateName`,
become reserved for the entire PodGroup. Previously, kube-scheduler could only
list individual Pods in a ResourceClaim's `status.reservedFor` field which is
limited to 256 items. Now, a single PodGroup reference in `status.reservedFor`
can represent many more than 256 Pods, allowing high-cardinality sharing of
devices.

Together, these changes enable massive workloads with complex topologies to
utilize DRA for scalable device management.

> ## Integration with the Job controller
> 
> To simplify the adoption of these advanced scheduling primitives, Kubernetes v1.36 introduces native integration between the Job controller
> and the Workload and PodGroup APIs. Without this integration, users or external orchestrators would have to manually manage the lifecycle
> of runtime PodGroup objects and carefully inject the proper references into every individual Pod template.
> Now, the Job controller automates this entire process.
> 
> When the [`WorkloadWithJob`](/docs/reference/command-line-tools-reference/feature-gates/#EnableWorkloadWithJob) feature gate is enabled,
> the Job controller automatically creates the necessary Workload and PodGroup objects and injects the `spec.schedulingGroup` field into
> all the Pods it creates. This automation triggers only when the Job spec meets the following constraints:
> 
> * `spec.parallelism` is greater than 1
> 
> * `spec.completionMode` is set to `Indexed`
> 
> * `spec.completions` is equal to `spec.parallelism`
> 
> * The `schedulingGroup` field is not already set on the Pod template
> 
> In future Kubernetes releases, this integration will expand to support additional workload controllers,
> and the current constraints for Jobs may be relaxed.

Reviewer 4: I suggest the following updates to this section, PTAL. 

```suggestion
## Integration with the Job controller

The Job controller can create and manage Workload and PodGroup objects on your behalf, so that Jobs representing a tightly coupled parallel application such as distributed training are gang-scheduled without any additional tooling. Before this integration, you had to either create the Workload and PodGroup yourself and wire their references into the Pod template, or rely on an external controller such as Kueue or JobSet to do it for you. Now, the Job controller automates this process natively.
When the [EnableWorkloadWithJob](https://github.com/docs/reference/command-line-tools-reference/feature-gates/#EnableWorkloadWithJob) feature gate is enabled, the Job controller automatically:

* creates a `Workload` and a corresponding runtime `PodGroup` for each
  qualifying Job,
* sets `spec.schedulingGroup` onto every Pod the Job creates so the
  scheduler treats them as a single gang, and
* sets the Job as the owner of the generated objects, so they are garbage
  collected when the Job is deleted.

### When does the integration kick in?

To keep the first iteration predictable, the Job controller only creates a
`Workload` and `PodGroup` when the Job has a well-defined, fixed shape:

* `.spec.parallelism` is greater than `1`
* [`spec.completionMode`](/docs/concepts/workloads/controllers/job/#completion-mode) is set to `Indexed`
* `.spec.completions` is equal to `.spec.parallelism`
* `.spec.template.spec.schedulingGroup` is not already set on the PodTemplate

These conditions describe the class of Jobs that gang scheduling can reason about: each Pod has a stable identity (Indexed), the gang size is known and fixed at admission time (parallelism == completions), and no other controller has already claimed scheduling responsibility (schedulingGroup filed is unset). Jobs that do not meet these conditions are scheduled pod-by-pod, exactly as before.

The generated Workload and PodGroup are owned by the Job through owner references, so they are garbage-collected when the Job is deleted. If you set `.spec.template.spec.schedulingGroup` yourself (for example, because a higher-level controller like Kueue or JobSet is managing the workload) the Job controller leaves the PodTemplate alone and does not create its own Workload or PodGroup. This makes the feature safe to enable in clusters that already use an external batch system. 

Here is an example of a Job that qualifies for gang scheduling:

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: training-job
  namespace: job-ns
spec:
  completionMode: Indexed
  parallelism: 4
  completions: 4
  template:
    spec:
      restartPolicy: Never
      containers:
      - name: worker
        image: registry.example/trainer:latest
 ```

The Job controller creates a `Workload` and a `PodGroup` owned by this Job, and every Pod it creates carries a `spec.schedulingGroup` that points at the generated `PodGroup`. The Pods are then held at a permit gate and placed together once all four can be scheduled at the same time by `PodGroup` scheduling cycle described earlier in this post.

### What's not covered yet

The current constraints limit this integration to static, indexed, fully-parallel Jobs. Support for additional workload shapes, including elastic Jobs and other built-in controllers, is tracked in [KEP-5547](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/5547-integrate-workload-with-job). 

In future Kubernetes releases, this integration will expand to support additional workload controllers, and the current constraints for Jobs may be relaxed.
```

Author: Thanks, applied!

`resourceClaimName` or the claim generated from `resourceClaimTemplateName`,
become reserved for the entire PodGroup. Previously, kube-scheduler could only
list individual Pods in a ResourceClaim's `status.reservedFor` field which is
limited to 256 items. Now, a single PodGroup reference in `status.reservedFor`
can represent many more than 256 Pods, allowing high-cardinality sharing of
devices.

Together, these changes enable massive workloads with complex topologies to
utilize DRA for scalable device management.

## Integration with the Job controller

In Kubernetes v1.36, the Job controller can create and manage Workload and PodGroup objects on your behalf,
so that Jobs representing a tightly coupled parallel application, such as distributed AI training,
are gang-scheduled without any additional tooling. Without this integration, you would have to
create the Workload and PodGroup yourself and wire their references into the Pod template.
Now, the Job controller automates this process natively.

When the [`WorkloadWithJob`](/docs/reference/command-line-tools-reference/feature-gates/#WorkloadWithJob)
feature gate is enabled, the Job controller automatically:

> * creates a `Workload` and a corresponding runtime `PodGroup` for each qualifying Job,

Reviewer: (nit / let's fix this after it merges)
```suggestion
* creates a Workload and a corresponding runtime PodGroup for each qualifying Job,
```

  feature gate on the `kube-scheduler`.
* Workload-aware preemption: Enable the
  [`WorkloadAwarePreemption`](/docs/reference/command-line-tools-reference/feature-gates/#WorkloadAwarePreemption)
  feature gate on the `kube-scheduler` (requires `GangScheduling` to also be enabled).
* DRA ResourceClaim support for workloads: Enable the
  [`DRAWorkloadResourceClaims`](/docs/reference/command-line-tools-reference/feature-gates/#DRAWorkloadResourceClaims)
  feature gate on the `kube-apiserver`, `kube-controller-manager`, `kube-scheduler` and `kubelet`.
* Workload API integration with the Job controller: Enable the
  [`WorkloadWithJob`](/docs/reference/command-line-tools-reference/feature-gates/#EnableWorkloadWithJob)
  feature gate on the `kube-apiserver` and `kube-controller-manager`.

We encourage you to try out workload-aware scheduling in your test clusters
and share your experiences to help shape the future of Kubernetes scheduling.
You can send your feedback by:

* Reaching out via [Slack (#workload-aware-scheduling)](https://kubernetes.slack.com/archives/C0AHLJ0EAEL).

> * Joining the [SIG Scheduling](https://docs.google.com/document/d/13mwye7nvrmV11q9_Eg77z-1w3X7Q1GTbslpml4J7F3A/edit)

Reviewer: nit: could link to https://www.kubernetes.dev/community/community-groups/sigs/scheduling/#meetings instead

  or [Workload API integration](https://docs.google.com/document/d/1XSPdK4L3zkAFhAZ3hBQJr2k7JX9CpGD7NeQfujM1PT4/edit) meetings.

> * Filing a new [issue](https://github.com/kubernetes/enhancements/issues) in the Kubernetes repository.

Reviewer: Let's **not** encourage people to do this; we don't want people to file KEPs without following the recommended process.

Author: Good catch, I wanted to link a k/k repo
